src/systemctl/systemctl.c

```diff
@@ -552,18 +553,45 @@ static int get_unit_list(
                         "org.freedesktop.systemd1",
                         "/org/freedesktop/systemd1",
                         "org.freedesktop.systemd1.Manager",
-                        "ListUnitsFiltered");
-
+                        "ListUnitsByStatesAndPatterns");
         if (r < 0)
                 return bus_log_create_error(r);
 
         r = sd_bus_message_append_strv(m, arg_states);
         if (r < 0)
                 return bus_log_create_error(r);
 
-        r = sd_bus_call(bus, m, 0, &error, &reply);
+        r = sd_bus_message_append_strv(m, patterns);
         if (r < 0)
-                return log_error_errno(r, "Failed to list units: %s", bus_error_message(&error, r));
+                return bus_log_create_error(r);
+
+        r = sd_bus_call(bus, m, 0, &error, &reply);
+        if (r < 0) {
+                // fallback to legacy ListUnitsFiltered method
+                fallback = true;
+                log_debug_errno(r, "Failed to list units: %s", bus_error_message(&error, r));
+                log_debug_errno(r, "Falling back to ListUnitsFiltered method.");
+                sd_bus_message_unrefp(&m);
```

Comment:
please use:

```c
m = sd_bus_message_unref(m);
```

The _unrefp() call is just a helper to make the `_cleanup_` logic nice to use. And m really should be assigned NULL here, so that it doesn't reference uninitialized memory which might break the error path if sd_bus_message_new_method_call() fails later on.